### ISSUE TITLE
Unable to bulk add with UUID PK in the presence of a TIMESTAMP column

### ISSUE THREAD

**Issue by sethgilchrist** [sqlalchemy/sqlalchemy#9822]
### Describe the bug

Bulk `add` in a `session` to a Postgres database using a string `UUID` PK does not work if the table definition contains a `TIMESTAMP` with a server-side default.

Two ways to circumvent the issue:
1. if the PK is passed as a `UUID` rather `string`, then a server-side `TIMESTAMP` is okay
2. if the `TIMESTAMP` is specified rather than being generated server-side, then using a `string` PK is okay.

### SQLAlchemy Version in Use

2.0.15

### DBAPI (i.e. the database driver)

psycopg2-binary Version: 2.9.6

### Database Vendor and Major Version

"PostgreSQL 12.13, compiled by Visual C++ build 1914, 64-bit"

### Python Version

Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] on win32

### Operating system

Windows and Linux

### To Reproduce

```python
from sqlalchemy import Column, String
from sqlalchemy.sql import func
from sqlalchemy.dialects.postgresql import UUID, TIMESTAMP
from sqlalchemy.orm import declarative_base, sessionmaker
from sqlalchemy import create_engine
from uuid import uuid4
from datetime import datetime

# define the database table
Base = declarative_base()


class my_object(Base):
    __tablename__ = 'my_object'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    val1 = Column(String)
    val2 = Column(String)
    time_created = Column(
        TIMESTAMP(timezone=True), default=func.current_timestamp(), nullable=False
    )


# connection information
server_host = 'localhost'
database_name = 'my_database'
username = 'my_user'
password = 'my_password'

engine = create_engine(f'postgresql://{username}:{password}@{server_host}/{database_name}')
Base.metadata.create_all(engine)
session_maker = sessionmaker(bind=engine)

# create some objects
objects_to_add = []
for i in range(10):
    new_object = {
        'id': str(uuid4()),
        # 'id': uuid4(),                                #  uncomment this line and it will work
        'val1': f'value {i}',
        'val2': f'value {i}',
        # 'time_created': datetime.now().astimezone()   #  uncomment this line and it will work
    }
    objects_to_add.append(new_object)

# add the objects to the database
with session_maker() as s:
    for obj in objects_to_add:
        s.add(my_object(**obj))
    s.commit()
```

### Error

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\default.py:891, in DefaultDialect._deliver_insertmanyvalues_batches(self, cursor, statement, parameters, generic_setinputsizes, context)
    889         else:
    890             # single-column sentinel with no value resolver
--> 891             ordered_rows = [
    892                 rows_by_sentinel[
    893                     parameters[_sentinel_param_key]  # type: ignore  # noqa: E501
    894                 ]
    895                 for parameters in imv_batch.batch
    896             ]
    897 except KeyError as ke:
    898     # see test_insert_exec.py::
    899     # IMVSentinelTest::test_sentinel_cant_match_keys
    900     # for coverage / demonstration

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\default.py:892, in (.0)
    889         else:
    890             # single-column sentinel with no value resolver
    891             ordered_rows = [
--> 892                 rows_by_sentinel[
    893                     parameters[_sentinel_param_key]  # type: ignore  # noqa: E501
    894                 ]
    895                 for parameters in imv_batch.batch
    896             ]
    897 except KeyError as ke:
    898     # see test_insert_exec.py::
    899     # IMVSentinelTest::test_sentinel_cant_match_keys
    900     # for coverage / demonstration

KeyError: '416a1d57-9591-4428-b728-1299220e2686'

The above exception was the direct cause of the following exception:

InvalidRequestError                       Traceback (most recent call last)
c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\test_bulk_add.py in line 43
     41 for obj in objects_to_add:
     42     s.add(my_object(**obj))
---> 43 s.commit()

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:1906, in Session.commit(self)
   1903 if trans is None:
   1904     trans = self._autobegin_t()
-> 1906 trans.commit(_to_root=True)

File :2, in commit(self, _to_root)

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\state_changes.py:137, in _StateChange.declare_states.._go(fn, self, *arg, **kw)
    135 self._next_state = _StateChangeStates.CHANGE_IN_PROGRESS
    136 try:
--> 137     ret_value = fn(self, *arg, **kw)
    138 except:
    139     raise

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:1221, in SessionTransaction.commit(self, _to_root)
   1219 if self._state is not SessionTransactionState.PREPARED:
   1220     with self._expect_state(SessionTransactionState.PREPARED):
-> 1221         self._prepare_impl()
   1223 if self._parent is None or self.nested:
   1224     for conn, trans, should_commit, autoclose in set(
   1225         self._connections.values()
   1226     ):

File :2, in _prepare_impl(self)

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\state_changes.py:137, in _StateChange.declare_states.._go(fn, self, *arg, **kw)
    135 self._next_state = _StateChangeStates.CHANGE_IN_PROGRESS
    136 try:
--> 137     ret_value = fn(self, *arg, **kw)
    138 except:
    139     raise

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:1196, in SessionTransaction._prepare_impl(self)
   1194     if self.session._is_clean():
   1195         break
-> 1196     self.session.flush()
   1197 else:
   1198     raise exc.FlushError(
   1199         "Over 100 subsequent flushes have occurred within "
   1200         "session.commit() - is an after_flush() hook "
   1201         "creating new objects?"
   1202     )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:4154, in Session.flush(self, objects)
   4152 try:
   4153     self._flushing = True
-> 4154     self._flush(objects)
   4155 finally:
   4156     self._flushing = False

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:4291, in Session._flush(self, objects)
   4289 except:
   4290     with util.safe_reraise():
-> 4291         transaction.rollback(_capture_exception=True)

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\util\langhelpers.py:147, in safe_reraise.__exit__(self, type_, value, traceback)
    145     assert exc_value is not None
    146     self._exc_info = None  # remove potential circular references
--> 147     raise exc_value.with_traceback(exc_tb)
    148 else:
    149     self._exc_info = None  # remove potential circular references

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\session.py:4251, in Session._flush(self, objects)
   4249 self._warn_on_events = True
   4250 try:
-> 4251     flush_context.execute()
   4252 finally:
   4253     self._warn_on_events = False

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\unitofwork.py:467, in UOWTransaction.execute(self)
    465 else:
    466     for rec in topological.sort(self.dependencies, postsort_actions):
--> 467         rec.execute(self)

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\unitofwork.py:644, in SaveUpdateAll.execute(self, uow)
    642 @util.preload_module("sqlalchemy.orm.persistence")
    643 def execute(self, uow):
--> 644     util.preloaded.orm_persistence.save_obj(
    645         self.mapper,
    646         uow.states_for_mapper_hierarchy(self.mapper, False, False),
    647         uow,
    648     )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\persistence.py:93, in save_obj(base_mapper, states, uowtransaction, single)
     81     update = _collect_update_commands(
     82         uowtransaction, table, states_to_update
     83     )
     85     _emit_update_statements(
     86         base_mapper,
     87         uowtransaction,
   (...)
     90         update,
     91     )
---> 93     _emit_insert_statements(
     94         base_mapper,
     95         uowtransaction,
     96         mapper,
     97         table,
     98         insert,
     99     )
    101 _finalize_insert_update_commands(
    102     base_mapper,
    103     uowtransaction,
   (...)
    119     ),
    120 )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\orm\persistence.py:1133, in _emit_insert_statements(base_mapper, uowtransaction, mapper, table, insert, bookkeeping, use_orm_insert_stmt, execution_options)
   1130 if do_executemany:
   1131     multiparams = [rec[2] for rec in records]
-> 1133     result = connection.execute(
   1134         statement, multiparams, execution_options=execution_options
   1135     )
   1137     if use_orm_insert_stmt is not None:
   1138         if return_result is None:

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\base.py:1413, in Connection.execute(self, statement, parameters, execution_options)
   1411     raise exc.ObjectNotExecutableError(statement) from err
   1412 else:
-> 1413     return meth(
   1414         self,
   1415         distilled_parameters,
   1416         execution_options or NO_OPTIONS,
   1417     )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\sql\elements.py:483, in ClauseElement._execute_on_connection(self, connection, distilled_params, execution_options)
    481     if TYPE_CHECKING:
    482         assert isinstance(self, Executable)
--> 483     return connection._execute_clauseelement(
    484         self, distilled_params, execution_options
    485     )
    486 else:
    487     raise exc.ObjectNotExecutableError(self)

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\base.py:1637, in Connection._execute_clauseelement(self, elem, distilled_parameters, execution_options)
   1625 compiled_cache: Optional[CompiledCacheType] = execution_options.get(
   1626     "compiled_cache", self.engine._compiled_cache
   1627 )
   1629 compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
   1630     dialect=dialect,
   1631     compiled_cache=compiled_cache,
   (...)
   1635     linting=self.dialect.compiler_linting | compiler.WARN_LINTING,
   1636 )
-> 1637 ret = self._execute_context(
   1638     dialect,
   1639     dialect.execution_ctx_cls._init_compiled,
   1640     compiled_sql,
   1641     distilled_parameters,
   1642     execution_options,
   1643     compiled_sql,
   1644     distilled_parameters,
   1645     elem,
   1646     extracted_params,
   1647     cache_hit=cache_hit,
   1648 )
   1649 if has_events:
   1650     self.dispatch.after_execute(
   1651         self,
   1652         elem,
   (...)
   1656         ret,
   1657     )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\base.py:1841, in Connection._execute_context(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)
   1838 context.pre_exec()
   1840 if context.execute_style is ExecuteStyle.INSERTMANYVALUES:
-> 1841     return self._exec_insertmany_context(
   1842         dialect,
   1843         context,
   1844     )
   1845 else:
   1846     return self._exec_single_context(
   1847         dialect, context, statement, parameters
   1848     )

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\base.py:2029, in Connection._exec_insertmany_context(self, dialect, context)
   2026 if self._echo:
   2027     stats = context._get_cache_stats() + " (insertmanyvalues)"
-> 2029 for imv_batch in dialect._deliver_insertmanyvalues_batches(
   2030     cursor,
   2031     str_statement,
   2032     effective_parameters,
   2033     generic_setinputsizes,
   2034     context,
   2035 ):
   2037     if imv_batch.processed_setinputsizes:
   2038         try:

File c:\Users\user\Documents\Python Projects\uuid_bulk_insert_error\.venv\lib\site-packages\sqlalchemy\engine\default.py:901, in DefaultDialect._deliver_insertmanyvalues_batches(self, cursor, statement, parameters, generic_setinputsizes, context)
    891                 ordered_rows = [
    892                     rows_by_sentinel[
    893                         parameters[_sentinel_param_key]  # type: ignore  # noqa: E501
    894                     ]
    895                     for parameters in imv_batch.batch
    896                 ]
    897     except KeyError as ke:
    898         # see test_insert_exec.py::
    899         # IMVSentinelTest::test_sentinel_cant_match_keys
    900         # for coverage / demonstration
--> 901         raise exc.InvalidRequestError(
    902             f"Can't match sentinel values in result set to "
    903             f"parameter sets; key {ke.args[0]!r} was not "
    904             "found. "
    905             "There may be a mismatch between the datatype "
    906             "passed to the DBAPI driver vs. that which it "
    907             "returns in a result row.  Try using a different "
    908             "datatype, such as integer"
    909         ) from ke
    911     result.extend(ordered_rows)
    913 else:

InvalidRequestError: Can't match sentinel values in result set to parameter sets; key '416a1d57-9591-4428-b728-1299220e2686' was not found. There may be a mismatch between the datatype passed to the DBAPI driver vs. that which it returns in a result row.  Try using a different datatype, such as integer
```

**Comment by zzzeek** [comment]
OK so, this confusion with UUID seems like it will keep coming up and we have to decide what we might do about it.

What I am curious about is if I point to these two things:

1. datatype is presented as ``UUID(as_uuid=True)``.   Note that as_uuid means [a Python uuid object is expected, not a string](https://docs.sqlalchemy.org/en/20/core/type_basics.html#sqlalchemy.types.UUID.params.as_uuid).  if we need to make that clearer, then that's one thing we can do.
2. actual data value is passed as a plain string.  Note that in Python, the objects ``'416a1d57-9591-4428-b728-1299220e2686'`` and ``uuid.UUID('416a1d57-9591-4428-b728-1299220e2686')`` are completely different, one is a string, one is a UUID. they do not compare as equal.

Just to put a point on it, the setting in 1. and the data in 2. don't match.   They "seem" like they work in some cases, but as we will see below, it actually never "works" completely when using it as a primary key with the ORM.

If we look at the error message, this is not an unexpected error, it's an actual case SQLAlchemy is testing:

    InvalidRequestError: Can't match sentinel values in result set to parameter sets; key '416a1d57-9591-4428-b728-1299220e2686' was not found. There may be a mismatch between the datatype passed to the DBAPI driver vs. that which it returns in a result row.  Try using a different datatype, such as integer


is it apparent what the problem is here? We can link this error to a new section in the errors.rst page if we need to put the context here.

The issue is that the database driver returns Python ``uuid.UUID()`` objects, not strings.  These need to match the data you put in for a column that's as critical as the primary key.   There's ways to make this coercion automatic, like using [TypeDecorator](https://docs.sqlalchemy.org/en/20/core/custom_types.html#typedecorator-recipes), but we would want to organize things such that people know what to do here.

your test "fails" both ways, even without the timestamp, just not as badly.  If we look at the situation described at #9573, even without the "INSERT" failure here, there is an identity map mismatch no matter what. Here's a demo of #9573 in terms of your program, no timestamp or error message:

```py
from uuid import uuid4

from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import sessionmaker

# define the database table
Base = declarative_base()


class my_object(Base):
    __tablename__ = 'my_object'

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    val1 = Column(String)
    val2 = Column(String)


engine = create_engine(f'postgresql://scott:tiger@localhost/test', echo='debug')
Base.metadata.create_all(engine)
session_maker = sessionmaker(bind=engine)



with session_maker() as s:
    mo = my_object(id=str(uuid4()))
    s.add(mo)
    s.flush()

    mo2 = s.scalars(select(my_object)).one()

    assert mo is mo2

```

the assertion at the end fails.   That's because if we look in the SQL output we see:

```
INSERT INTO my_object (id, val1, val2) VALUES (%(id)s::UUID, %(val1)s, %(val2)s)
2023-05-22 18:57:54,347 INFO sqlalchemy.engine.Engine [generated in 0.00017s] {'id': '99f01b06-afad-4afa-8197-524abfef760c', 'val1': None, 'val2': None}
2023-05-22 18:57:54,349 INFO sqlalchemy.engine.Engine SELECT my_object.id, my_object.val1, my_object.val2 
FROM my_object
2023-05-22 18:57:54,349 INFO sqlalchemy.engine.Engine [generated in 0.00008s] {}
2023-05-22 18:57:54,349 DEBUG sqlalchemy.engine.Engine Col ('id', 'val1', 'val2')
2023-05-22 18:57:54,349 DEBUG sqlalchemy.engine.Engine Row (UUID('99f01b06-afad-4afa-8197-524abfef760c'), None, None)

```

String goes in, driver returns UUID().  Different object goes in the identity map.

So what we need to figure out here is do we want to do something like add a new parameter on ``UUID`` that does "expect_strings_or_uuids"....we dont assert this by default because `isinstance()` and coercion checks add measurable performance expense, and we dont want to by default add performance expense to all applications, including the majority of them which pass around ``UUID()`` objects when `as_uuid=True`.

basically, like in #9573, I don't have a good answer for this use case of sending different datatypes than those requested, without putting expensive checks and coercions everywhere, which we've historically not done.  UUID is a place where this will happen since there are two forms of UUID; that is, we dont get this so much with like int/float, or datetime/string date.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**attempt to clarify sentinel mismatch message** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4628

**Comment by zzzeek** [comment]
there's really no way to "fix" the version of this that's #9573 actually without using an ORM validates rule that changes the datatype at the object level to match what would be returned by the database.

**Comment by sethgilchrist** [comment]
Thanks for the thorough response and explanation.  I think the critical lines from #9573 is 

> ...ORM objects expect a fully consistent Python type for attributes whether they are set from the Python side or they are delivered from the database side.
>...types will then include a "result processor" that converts back to the Python-side type. So the Python side sees only one datatype, the database sees only another

My expectation was that setting `as_uuid=True` would return UUIDs and optionally accept UUIDs from the Python side.  Basically it would allow the user to use UUID on the python side rather than requiring strings.  I understand now.
